python/ray/workflow/tests/test_checkpoint.py

```diff
@@ -0,0 +1,164 @@
+import ray
+import time
+import pytest
+from ray.tests.conftest import *  # noqa
+
+import numpy as np
+from ray import workflow
+from ray.workflow.tests import utils
+from ray.exceptions import RaySystemError
+
+
+@workflow.step
+def large_input():
+    return np.arange(2**24)
+
+
+@workflow.step
+def identity(x):
+    return x
+
+
+@workflow.step
+def average(x):
+    return np.mean(x)
+
+
+@workflow.step
+def checkpoint_dag(checkpoint):
+    x = large_input.options(checkpoint=checkpoint).step()
+    y = identity.options(checkpoint=checkpoint).step(x)
+    return average.step(y)
+
+
+def test_checkpoint_dag(workflow_start_regular):
```

Comment:
It would be nice if the tests didn't depend on run time, which is a bit unpredictable in CI.

Could we design a test so that we check whether a task is rerun or not after a workflow failure? We could use an external actor to count how many times the task is run.

It would also be nice to check the case that I mentioned above:
```
grandparent(checkpoint=True):
  parent(checkpoint=False):
    child(checkpoint=True)
```